File: src/ua/edu/hneu/oop/java/lr5/task2/StudentFactory.java
package ua.edu.hneu.oop.java.lr5.task2;

import ua.edu.hneu.oop.java.util.ConsoleUtil;

import java.io.IOException;
import java.util.Arrays;
import java.util.Objects;
import java.util.function.Predicate;

public class StudentFactory {
    private static final String SURNAME_REGEX = "(?u)\\p{L}+ *";
    private static final String INITIALS_REGEX = "(\\p{L}+\\.)( ?\\p{L}+\\.)? *";
    private static final String ARRAY_DELIMITERS = "[;,:]";
    private static final String INT_ARRAY_REGEX = "\\d *(" + ARRAY_DELIMITERS + " *\\d)* *";

    private static final String SURNAME_INVITATION = "Enter student's surname:";
    private static final String INITIALS_INVITATION = "Enter student's initials:";
    private static final String GROUP_INVITATION = "Enter student's group number:";
    private static final String MARKS_INVITATION = "Enter student's marks in format \"4,5,3,2,4\":";
    private static final String WRONG_INPUT = "Wrong input. Please try again." + System.lineSeparator();

    private ConsoleUtil consoleUtill;

    public StudentFactory(ConsoleUtil consoleUtil) {
        this.consoleUtill = consoleUtil;
    }

    public Student createStudent() throws IOException {
        consoleUtill.println(SURNAME_INVITATION);
        String surname = obtainString(this::validateSurname);
        consoleUtill.println(INITIALS_INVITATION);
        String initials = obtainString(this::validateInitials);
        consoleUtill.println(GROUP_INVITATION);
        int group = obtainInt(this::validateGroup);
        consoleUtill.println(MARKS_INVITATION);
        int[] marks = obtainIntArr();
        return new Student(surname, initials, group, marks);
    }

    private String obtainString(Predicate<String> validator) throws IOException {
        do {
            String input = consoleUtill.readInput();
            if (validator.test(input)) {
                return input.trim();
            }
            consoleUtill.print(WRONG_INPUT);
        } while (true);
    }

    private int obtainInt(Predicate<Integer> validator) throws IOException {
        do {
            int input = Integer.parseInt(consoleUtill.readInput());
            if (validator.test(input)) {
                return input;
            }
            consoleUtill.print(WRONG_INPUT);
        } while (true);
    }

    private int[] obtainIntArr() throws IOException {
        do {
            String input = consoleUtill.readInput();
            if (validateMarksArrayString(input)) {
                int[] arr = Arrays.stream(input.split(ARRAY_DELIMITERS)).map(String::trim).mapToInt(Integer::parseInt).toArray();
                if (arr.length == Student.MARKS_SIZE) {
                    return arr;
                }
            }
            consoleUtill.print(WRONG_INPUT);
        } while (true);
    }

    private boolean validateSurname(String surname) {
        return Objects.nonNull(surname) &&
                surname.matches(SURNAME_REGEX);
    }

    private boolean validateInitials(String initials) {
        return Objects.nonNull(initials) &&
                initials.matches(INITIALS_REGEX);
    }

    private boolean validateGroup(int group) {
        return group > 0;
    }

    private boolean validateMarksArrayString(String input) {
        return Objects.nonNull(input) &&
                input.matches(INT_ARRAY_REGEX);
    }
}

File: src/ua/edu/hneu/oop/dotNet/LR5/Task2/StudentFactory .cs
using System;
using System.Linq;
using System.Text.RegularExpressions;

namespace Task2
{
    public class StudentFactory
    {
        private const string SurnameRegex = "\\w";
        private const string InitialsRegex = "\\w\\.\\w*";
        private const string ArrayDelimiters = "[;,:]";
        private const string IntArrayRegex = "\\d *(" + ArrayDelimiters + " *\\d)* *";

        private const string SurnameInvitation = "Enter student's surname:";
        private const string InitialsInvitation = "Enter student's initials:";
        private const string GroupInvitation = "Enter student's group number:";
        private const string MarksInvitation = "Enter student's marks in format \"4,5,3,2,4\":";
        private const string WrongInput = "Wrong input. Please try again.\n";

        public Student CreateStudent()
        {
            Console.WriteLine(SurnameInvitation);
            String surname = GetString(validateSurname);
            Console.WriteLine(InitialsInvitation);
            String initials = GetString(ValidateInitials);
            Console.WriteLine(GroupInvitation);
            int group = GetGroup(ValidateGroup);
            Console.WriteLine(MarksInvitation);

            int[] marks = GetMarks();
            return new Student(surname, initials, group, marks);
        }

        private String GetString(Predicate<String> validator)
        {
            do
            {
                String input = Console.ReadLine();

                if (validator(input))
                    return input.Trim();

                Console.WriteLine(WrongInput);
            } while (true);
        }

        private int GetGroup(Predicate<int> validator)
        {
            do
            {
                string s = Console.ReadLine();

                if (int.TryParse(s, out int input) && validator(input))
                    return input;

                Console.WriteLine(WrongInput);
            } while (true);
        }

        private int[] GetMarks()
        {
            do
            {
                String input = Console.ReadLine();
                if (ValidateMarksArrayString(input))
                {
                    int[] arr = input.Split(ArrayDelimiters.Substring(1,3)
                        .ToCharArray())
                        .Select(x => x.Trim())
                        .Select(int.Parse)
                        .ToArray();

                    if (arr.Length == Student.MarksSize)
                        return arr;
                }
                Console.WriteLine(WrongInput);
            } while (true);
        }

        private bool validateSurname(String surname) => !string.IsNullOrEmpty(surname) && Regex.IsMatch(surname, SurnameRegex);

        private bool ValidateInitials(String initials) => !string.IsNullOrEmpty(initials) && Regex.IsMatch(initials, InitialsRegex);

        private bool ValidateGroup(int group) => group > 0;

        private bool ValidateMarksArrayString(String input) => !string.IsNullOrEmpty(input) && Regex.IsMatch(input, IntArrayRegex);
    }
}